File: mobile/android/src/main/java/org/intcoin/sdk/INTcoinSDK.kt
// Copyright (c) 2024-2025 The INTcoin Core developers
// Distributed under the MIT software license

package org.intcoin.sdk

import android.content.Context
import java.io.File

/**
 * INTcoin Mobile SDK for Android
 * Provides native Kotlin interface for INTcoin lightweight wallet functionality
 */
class INTcoinSDK private constructor(
    private val network: String,
    private val walletPath: String,
    private val rpcEndpoint: String
) : AutoCloseable {

    private var sdkHandle: Long = 0
    private var transactionCallback: ((TransactionEvent) -> Unit)? = null
    private var syncProgressCallback: ((SyncProgress) -> Unit)? = null

    init {
        // Load native library
        System.loadLibrary("intcoin_core")

        // Initialize SDK
        sdkHandle = nativeCreate(network, walletPath, rpcEndpoint)
        if (sdkHandle == 0L) {
            throw INTcoinException("Failed to initialize INTcoin SDK")
        }
    }

    override fun close() {
        if (sdkHandle != 0L) {
            nativeDestroy(sdkHandle)
            sdkHandle = 0
        }
    }

    // MARK: - Wallet Management

    /**
     * Create new wallet with password encryption
     * @param password Wallet encryption password
     * @return BIP39 mnemonic phrase (store safely!)
     */
    @Throws(INTcoinException::class)
    fun createWallet(password: String): String {
        checkHandle()
        return nativeCreateWallet(sdkHandle, password)
            ?: throw INTcoinException("Failed to create wallet")
    }

    /**
     * Open existing wallet
     * @param password Wallet password
     */
    @Throws(INTcoinException::class)
    fun openWallet(password: String) {
        checkHandle()
        if (!nativeOpenWallet(sdkHandle, password)) {
            throw INTcoinException("Failed to open wallet - incorrect password?")
        }
    }

    /**
     * Close wallet
     */
    fun closeWallet() {
        if (sdkHandle != 0L) {
            nativeCloseWallet(sdkHandle)
        }
    }

    // MARK: - Address Management

    /**
     * Generate new receiving address
     * @return INTcoin Bech32 address
     */
    @Throws(INTcoinException::class)
    fun getNewAddress(): String {
        checkHandle()
        return nativeGetNewAddress(sdkHandle)
            ?: throw INTcoinException("Failed to generate address")
    }

    // MARK: - Balance & Transactions

    /**
     * Get wallet balance
     * @return Wallet balance in INTS
     */
    @Throws(INTcoinException::class)
    fun getBalance(): Balance {
        checkHandle()
        return nativeGetBalance(sdkHandle)
            ?: throw INTcoinException("Failed to query balance")
    }

    /**
     * Send transaction
     * @param toAddress Recipient address
     * @param amountINTS Amount in INTS (1 INT = 1,000,000 INTS)
     * @return Transaction hash as byte array
     */
    @Throws(INTcoinException::class)
    fun sendTransaction(toAddress: String, amountINTS: Long): ByteArray {
        checkHandle()
        if (!validateAddress(toAddress)) {
            throw INTcoinException("Invalid recipient address")
        }
        return nativeSendTransaction(sdkHandle, toAddress, amountINTS)
            ?: throw INTcoinException("Failed to send transaction")
    }

    // MARK: - Sync & Network

    /**
     * Start blockchain sync
     */
    @Throws(INTcoinException::class)
    fun startSync() {
        checkHandle()
        if (!nativeStartSync(sdkHandle)) {
            throw INTcoinException("Failed to start sync")
        }
    }

    /**
     * Stop blockchain sync
     */
    fun stopSync() {
        if (sdkHandle != 0L) {
            nativeStopSync(sdkHandle)
        }
    }

    /**
     * Get sync progress (0.0 to 1.0)
     * @return Sync progress as fraction
     */
    fun getSyncProgress(): Double {
        return if (sdkHandle != 0L) {
            nativeGetSyncProgress(sdkHandle)
        } else {
            0.0
        }
    }

    // MARK: - Callbacks

    /**
     * Set transaction event callback
     * @param callback Function to call on transaction events
     */
    fun setTransactionCallback(callback: (TransactionEvent) -> Unit) {
        this.transactionCallback = callback
    }

    /**
     * Set sync progress callback
     * @param callback Function to call on sync progress updates
     */
    fun setSyncProgressCallback(callback: (SyncProgress) -> Unit) {
        this.syncProgressCallback = callback
    }

    // MARK: - Private Methods

    private fun checkHandle() {
        if (sdkHandle == 0L) {
            throw INTcoinException("SDK not initialized")
        }
    }

    // MARK: - Native Methods

    private external fun nativeCreate(network: String, walletPath: String, rpcEndpoint: String): Long
    private external fun nativeDestroy(handle: Long)
    private external fun nativeCreateWallet(handle: Long, password: String): String?
    private external fun nativeOpenWallet(handle: Long, password: String): Boolean
    private external fun nativeCloseWallet(handle: Long)
    private external fun nativeGetNewAddress(handle: Long): String?
    private external fun nativeGetBalance(handle: Long): Balance?
    private external fun nativeSendTransaction(handle: Long, toAddress: String, amountINTS: Long): ByteArray?
    private external fun nativeStartSync(handle: Long): Boolean
    private external fun nativeStopSync(handle: Long)
    private external fun nativeGetSyncProgress(handle: Long): Double

    companion object {
        /**
         * Create SDK instance
         * @param context Android context
         * @param network Network type ("mainnet" or "testnet")
         * @param rpcEndpoint RPC endpoint URL
         * @return SDK instance
         */
        @JvmStatic
        fun create(
            context: Context,
            network: String = "mainnet",
            rpcEndpoint: String = "http://localhost:2210"
        ): INTcoinSDK {
            // Use app's files directory for wallet storage
            val walletPath = File(context.filesDir, "intcoin_wallet").absolutePath

            return INTcoinSDK(network, walletPath, rpcEndpoint)
        }

        /**
         * Validate INTcoin address format
         * @param address Address to validate
         * @return True if valid
         */
        @JvmStatic
        fun validateAddress(address: String): Boolean {
            return nativeValidateAddress(address)
        }

        /**
         * Format INTS to human-readable string
         * @param ints Amount in INTS
         * @return Formatted string (e.g., "1.234567 INT")
         */
        @JvmStatic
        fun formatINTS(ints: Long): String {
            return nativeFormatINTS(ints)
        }

        /**
         * Parse INT amount string to INTS
         * @param amountString Amount string (e.g., "1.5" or "1500000")
         * @return Amount in INTS, or null if invalid
         */
        @JvmStatic
        fun parseINTAmount(amountString: String): Long? {
            val dotIndex = amountString.indexOf('.')

            return if (dotIndex == -1) {
                // No decimal point, assume INTS
                amountString.toLongOrNull()
            } else {
                // Has decimal point, parse as INT
                try {
                    val intPart = amountString.substring(0, dotIndex)
                    var fracPart = amountString.substring(dotIndex + 1)

                    // Pad fraction to 6 digits
                    while (fracPart.length < 6) {
                        fracPart += "0"
                    }
                    if (fracPart.length > 6) {
                        fracPart = fracPart.substring(0, 6)
                    }

                    val intValue = intPart.toLong()
                    val fracValue = fracPart.toLong()

                    intValue * 1_000_000 + fracValue
                } catch (e: Exception) {
                    null
                }
            }
        }

        /**
         * Generate payment URI for QR code
         * @param address Receiving address
         * @param amountINTS Amount in INTS (0 for no amount)
         * @param label Payment label (optional)
         * @param message Payment message (optional)
         * @return intcoin: URI for QR code generation
         */
        @JvmStatic
        fun generatePaymentURI(
            address: String,
            amountINTS: Long = 0,
            label: String? = null,
            message: String? = null
        ): String {
            return nativeGeneratePaymentURI(address, amountINTS, label, message)
        }

        // Native companion methods
        @JvmStatic
        private external fun nativeValidateAddress(address: String): Boolean

        @JvmStatic
        private external fun nativeFormatINTS(ints: Long): String

        @JvmStatic
        private external fun nativeGeneratePaymentURI(
            address: String,
            amountINTS: Long,
            label: String?,
            message: String?
        ): String
    }
}

// MARK: - Data Classes

/**
 * Wallet balance
 */
data class Balance(
    /** Confirmed balance in INTS */
    val confirmed: Long,

    /** Unconfirmed balance in INTS */
    val unconfirmed: Long
) {
    /** Total balance (confirmed + unconfirmed) in INTS */
    val total: Long
        get() = confirmed + unconfirmed

    /** Confirmed balance as formatted string */
    val confirmedFormatted: String
        get() = INTcoinSDK.formatINTS(confirmed)

    /** Unconfirmed balance as formatted string */
    val unconfirmedFormatted: String
        get() = INTcoinSDK.formatINTS(unconfirmed)

    /** Total balance as formatted string */
    val totalFormatted: String
        get() = INTcoinSDK.formatINTS(total)
}

/**
 * Transaction event
 */
data class TransactionEvent(
    val type: Type,
    val txHash: ByteArray,
    val address: String,
    val amountINTS: Long,
    val confirmations: Int,
    val timestamp: Long
) {
    enum class Type {
        RECEIVED, SENT, CONFIRMED, PENDING
    }

    val amountFormatted: String
        get() = INTcoinSDK.formatINTS(amountINTS)

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as TransactionEvent

        if (!txHash.contentEquals(other.txHash)) return false

        return true
    }

    override fun hashCode(): Int {
        return txHash.contentHashCode()
    }
}

/**
 * Sync progress
 */
data class SyncProgress(
    val currentHeight: Long,
    val targetHeight: Long,
    val progress: Double,  // 0.0 to 1.0
    val isSyncing: Boolean
) {
    /** Progress as percentage (0-100) */
    val percentage: Double
        get() = progress * 100.0
}

/**
 * INTcoin SDK exception
 */
class INTcoinException(message: String) : Exception(message)

// MARK: - Extensions

/**
 * Convert byte array to hex string
 */
fun ByteArray.toHex(): String {
    return joinToString("") { "%02x".format(it) }
}

File: mobile/ios/INTcoinSDK.swift
// Copyright (c) 2024-2025 The INTcoin Core developers
// Distributed under the MIT software license

import Foundation

/// INTcoin Mobile SDK for iOS
/// Provides native Swift interface for INTcoin lightweight wallet functionality
public class INTcoinSDK {

    // MARK: - Properties

    private var sdkHandle: OpaquePointer?
    private let network: String
    private let walletPath: String
    private let rpcEndpoint: String

    // Callbacks
    private var transactionCallback: ((TransactionEvent) -> Void)?
    private var syncProgressCallback: ((SyncProgress) -> Void)?

    // MARK: - Initialization

    /// Initialize INTcoin SDK
    /// - Parameters:
    ///   - network: Network type ("mainnet" or "testnet")
    ///   - walletPath: Path to wallet storage directory
    ///   - rpcEndpoint: RPC endpoint URL
    public init(network: String = "mainnet",
                walletPath: String? = nil,
                rpcEndpoint: String = "http://localhost:2210") throws {

        self.network = network
        self.rpcEndpoint = rpcEndpoint

        // Use app's documents directory if no path provided
        if let path = walletPath {
            self.walletPath = path
        } else {
            let documentsPath = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
            self.walletPath = documentsPath.appendingPathComponent("intcoin_wallet").path
        }

        // Create SDK instance
        sdkHandle = intcoin_sdk_create(
            network.cString(using: .utf8),
            self.walletPath.cString(using: .utf8),
            rpcEndpoint.cString(using: .utf8)
        )

        guard sdkHandle != nil else {
            throw INTcoinError.initializationFailed
        }
    }

    deinit {
        if let handle = sdkHandle {
            intcoin_sdk_destroy(handle)
        }
    }

    // MARK: - Wallet Management

    /// Create new wallet with password encryption
    /// - Parameter password: Wallet encryption password
    /// - Returns: BIP39 mnemonic phrase (store safely!)
    public func createWallet(password: String) throws -> String {
        guard let handle = sdkHandle else {
            throw INTcoinError.sdkNotInitialized
        }

        var mnemonicBuffer = [CChar](repeating: 0, count: 256)
        let result = intcoin_sdk_create_wallet(
            handle,
            password.cString(using: .utf8),
            &mnemonicBuffer
        )

        guard result == 0 else {
            throw INTcoinError.walletCreationFailed
        }

        return String(cString: mnemonicBuffer)
    }

    /// Open existing wallet
    /// - Parameter password: Wallet password
    public func openWallet(password: String) throws {
        guard let handle = sdkHandle else {
            throw INTcoinError.sdkNotInitialized
        }

        let result = intcoin_sdk_open_wallet(handle, password.cString(using: .utf8))

        guard result == 0 else {
            throw INTcoinError.walletOpenFailed
        }
    }

    /// Close wallet
    public func closeWallet() {
        guard let handle = sdkHandle else { return }
        intcoin_sdk_close_wallet(handle)
    }

    // MARK: - Address Management

    /// Generate new receiving address
    /// - Returns: INTcoin Bech32 address
    public func getNewAddress() throws -> String {
        guard let handle = sdkHandle else {
            throw INTcoinError.sdkNotInitialized
        }

        var addressBuffer = [CChar](repeating: 0, count: 64)
        let result = intcoin_sdk_get_new_address(handle, &addressBuffer)

        guard result == 0 else {
            throw INTcoinError.addressGenerationFailed
        }

        return String(cString: addressBuffer)
    }

    /// Validate INTcoin address format
    /// - Parameter address: Address to validate
    /// - Returns: True if valid
    public static func validateAddress(_ address: String) -> Bool {
        return intcoin_sdk_validate_address(address.cString(using: .utf8)) == 1
    }

    // MARK: - Balance & Transactions

    /// Get wallet balance
    /// - Returns: Wallet balance in INTS
    public func getBalance() throws -> Balance {
        guard let handle = sdkHandle else {
            throw INTcoinError.sdkNotInitialized
        }

        var confirmed: UInt64 = 0
        var unconfirmed: UInt64 = 0
        let result = intcoin_sdk_get_balance(handle, &confirmed, &unconfirmed)

        guard result == 0 else {
            throw INTcoinError.balanceQueryFailed
        }

        return Balance(confirmed: confirmed, unconfirmed: unconfirmed)
    }

    /// Send transaction
    /// - Parameters:
    ///   - toAddress: Recipient address
    ///   - amountINTS: Amount in INTS (1 INT = 1,000,000 INTS)
    /// - Returns: Transaction hash
    public func sendTransaction(toAddress: String, amountINTS: UInt64) throws -> Data {
        guard let handle = sdkHandle else {
            throw INTcoinError.sdkNotInitialized
        }

        var txHashBuffer = [UInt8](repeating: 0, count: 32)
        let result = intcoin_sdk_send_transaction(
            handle,
            toAddress.cString(using: .utf8),
            amountINTS,
            &txHashBuffer
        )

        guard result == 0 else {
            throw INTcoinError.transactionSendFailed
        }

        return Data(txHashBuffer)
    }

    // MARK: - Sync & Network

    /// Start blockchain sync
    public func startSync() throws {
        guard let handle = sdkHandle else {
            throw INTcoinError.sdkNotInitialized
        }

        let result = intcoin_sdk_start_sync(handle)

        guard result == 0 else {
            throw INTcoinError.syncStartFailed
        }
    }

    /// Stop blockchain sync
    public func stopSync() {
        guard let handle = sdkHandle else { return }
        intcoin_sdk_stop_sync(handle)
    }

    /// Get sync progress (0.0 to 1.0)
    /// - Returns: Sync progress as percentage
    public func getSyncProgress() -> Double {
        guard let handle = sdkHandle else { return 0.0 }
        return intcoin_sdk_get_sync_progress(handle)
    }

    // MARK: - QR Code Support

    /// Generate payment URI for QR code
    /// - Parameters:
    ///   - address: Receiving address
    ///   - amountINTS: Amount in INTS (optional, 0 for no amount)
    ///   - label: Payment label (optional)
    ///   - message: Payment message (optional)
    /// - Returns: intcoin: URI for QR code generation
    public static func generatePaymentURI(address: String,
                                         amountINTS: UInt64 = 0,
                                         label: String? = nil,
                                         message: String? = nil) -> String {
        var uriBuffer = [CChar](repeating: 0, count: 512)
        intcoin_sdk_generate_payment_uri(
            address.cString(using: .utf8),
            amountINTS,
            label?.cString(using: .utf8),
            message?.cString(using: .utf8),
            &uriBuffer
        )
        return String(cString: uriBuffer)
    }

    // MARK: - Utility

    /// Format INTS to human-readable string
    /// - Parameter ints: Amount in INTS
    /// - Returns: Formatted string (e.g., "1.234567 INT")
    public static func formatINTS(_ ints: UInt64) -> String {
        var buffer = [CChar](repeating: 0, count: 32)
        intcoin_sdk_format_ints(ints, &buffer)
        return String(cString: buffer)
    }

    /// Parse INT amount string to INTS
    /// - Parameter amountString: Amount string (e.g., "1.5" or "1500000")
    /// - Returns: Amount in INTS
    public static func parseINTAmount(_ amountString: String) -> UInt64? {
        // Parse "1.5" or "1500000" to INTS
        if let dotIndex = amountString.firstIndex(of: ".") {
            // Has decimal point, parse as INT
            let intPart = String(amountString[..<dotIndex])
            var fracPart = String(amountString[amountString.index(after: dotIndex)...])

            // Pad fraction to 6 digits
            while fracPart.count < 6 {
                fracPart += "0"
            }
            if fracPart.count > 6 {
                fracPart = String(fracPart.prefix(6))
            }

            guard let intValue = UInt64(intPart), let fracValue = UInt64(fracPart) else {
                return nil
            }

            return intValue * 1_000_000 + fracValue
        } else {
            // No decimal point, assume INTS
            return UInt64(amountString)
        }
    }

    // MARK: - Callbacks

    /// Set transaction event callback
    /// - Parameter callback: Callback function for transaction events
    public func setTransactionCallback(_ callback: @escaping (TransactionEvent) -> Void) {
        self.transactionCallback = callback
    }

    /// Set sync progress callback
    /// - Parameter callback: Callback function for sync progress updates
    public func setSyncProgressCallback(_ callback: @escaping (SyncProgress) -> Void) {
        self.syncProgressCallback = callback
    }
}

// MARK: - Data Types

/// Wallet balance
public struct Balance {
    /// Confirmed balance in INTS
    public let confirmed: UInt64

    /// Unconfirmed balance in INTS
    public let unconfirmed: UInt64

    /// Total balance (confirmed + unconfirmed) in INTS
    public var total: UInt64 {
        return confirmed + unconfirmed
    }

    /// Confirmed balance as formatted string
    public var confirmedFormatted: String {
        return INTcoinSDK.formatINTS(confirmed)
    }

    /// Unconfirmed balance as formatted string
    public var unconfirmedFormatted: String {
        return INTcoinSDK.formatINTS(unconfirmed)
    }

    /// Total balance as formatted string
    public var totalFormatted: String {
        return INTcoinSDK.formatINTS(total)
    }
}

/// Transaction event
public struct TransactionEvent {
    public enum EventType {
        case received, sent, confirmed, pending
    }

    public let type: EventType
    public let txHash: Data
    public let address: String
    public let amountINTS: UInt64
    public let confirmations: UInt32
    public let timestamp: Date
}

/// Sync progress
public struct SyncProgress {
    public let currentHeight: UInt64
    public let targetHeight: UInt64
    public let progress: Double  // 0.0 to 1.0
    public let isSyncing: Bool

    /// Progress as percentage (0-100)
    public var percentage: Double {
        return progress * 100.0
    }
}

/// INTcoin SDK errors
public enum INTcoinError: Error {
    case initializationFailed
    case sdkNotInitialized
    case walletCreationFailed
    case walletOpenFailed
    case addressGenerationFailed
    case balanceQueryFailed
    case transactionSendFailed
    case syncStartFailed
    case invalidAddress
    case insufficientBalance

    public var localizedDescription: String {
        switch self {
        case .initializationFailed:
            return "Failed to initialize INTcoin SDK"
        case .sdkNotInitialized:
            return "SDK not initialized"
        case .walletCreationFailed:
            return "Failed to create wallet"
        case .walletOpenFailed:
            return "Failed to open wallet"
        case .addressGenerationFailed:
            return "Failed to generate address"
        case .balanceQueryFailed:
            return "Failed to query balance"
        case .transactionSendFailed:
            return "Failed to send transaction"
        case .syncStartFailed:
            return "Failed to start sync"
        case .invalidAddress:
            return "Invalid INTcoin address"
        case .insufficientBalance:
            return "Insufficient balance"
        }
    }
}
